src/modules/update/common/update-common.c

```diff
@@ -0,0 +1,488 @@
+    }
+
+    /* If some member is empty, give up */
+    if (!(response.url && response.version
+        && response.hash && response.hash_algorithm)) {
+        SOL_WRN("Malformed check response");
+        handle->cb_check((void *)handle->user_data, -EINVAL, NULL);
+        free_response_members(&response);
+        delete_handle(handle);
+        return;
+    }
+
+    handle->cb_check((void *)handle->user_data, 0, &response);
+    free_response_members(&response);
+    delete_handle(handle);
+
+    return;
+}
+
+static void
+on_digest_ready_cb(void *data, struct sol_message_digest *md, struct sol_blob *output)
+{
+    struct sol_update_handle *handle = data;
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+    struct sol_str_slice slice = sol_str_slice_from_blob(output);
+    int r = 0;
+
+    r = sol_buffer_append_as_base16(&buffer, slice, false);
+    SOL_INT_CHECK_GOTO(r, < 0, end);
+
+    if (buffer.used != strlen(handle->hash) ||
+        strncmp(buffer.data, handle->hash, buffer.used != 0)) {
+
+        r = -EINVAL;
+        SOL_WRN("Expected hash differs of download file hash, expected [%s], found [%.*s]",
+            handle->hash, (int)buffer.used, (char *)buffer.data);
+    }
+
+end:
+    handle->cb_hash(handle, r);
+    sol_message_digest_del(md);
+    sol_buffer_fini(&buffer);
+}
+
+static void
+on_feed_done_cb(void *data, struct sol_message_digest *md, struct sol_blob *input)
+{
+    struct sol_update_handle *handle = data;
+    char buf[CHUNK_SIZE], *blob_backend = NULL;
+    struct sol_blob *blob = NULL;
+    size_t size;
+    bool last;
+    int r;
+
+    size = fread(buf, 1, sizeof(buf), handle->file);
+    if (ferror(handle->file)) {
+        SOL_WRN("Could not read file for feed hash algorithm");
+        goto err;
+    }
+
+    last = feof(handle->file);
+
+    /* TODO Maybe this is a bug on sol_message_digest? Keeps calling on_feed_done
+     * after send last chunk */
+    if (!size && last) {
+        SOL_WRN("Nothing more to feed hash algorithm, ignoring on_feed_done request");
+        return;
+    }
+
+    blob_backend = malloc(size);
+    SOL_NULL_CHECK_GOTO(blob_backend, err);
+
+    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, blob_backend, size);
+    SOL_NULL_CHECK_GOTO(blob, err);
+
+    memcpy(blob_backend, buf, size);
+
+    r = sol_message_digest_feed(md, blob, last);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    sol_blob_unref(blob);
+    return;
+
+err:
+    SOL_WRN("Could not feed data to check update file hash");
+    free(blob_backend);
+    sol_blob_unref(blob);
+    sol_message_digest_del(md);
+    handle->cb_hash(handle, -EINVAL);
+}
+
+static bool
+check_hash(struct sol_update_handle *handle,
+    void (*cb)(struct sol_update_handle *handle, int status))
+{
+    struct sol_message_digest_config cfg = {
+        SOL_SET_API_VERSION(.api_version = SOL_MESSAGE_DIGEST_CONFIG_API_VERSION, )
+        .algorithm = handle->hash_algorithm,
```

Comment:
On a serious side, depends on backend implementation. But the reference one might expect for some specific ones.